connectors/sources/microsoft_teams.py

```diff
@@ -0,0 +1,360 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""Microsoft Teams source module responsible to fetch documents from Microsoft Teams.
+"""
+from datetime import datetime, timedelta
+from enum import Enum
+from functools import cached_property
+
+import aiohttp
+from aiohttp.client_exceptions import ClientResponseError
+from msal import ConfidentialClientApplication
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource
+from connectors.utils import (
+    CacheWithTimeout,
+    CancellableSleeps,
+    RetryStrategy,
+    iso_utc,
+    retryable,
+)
+
+MAX_FILE_SIZE = 10485760
+TOKEN_EXPIRES = 3599
+RETRY_COUNT = 3
+RETRY_INTERVAL = 2
+GRAPH_API_AUTH_URL = "https://login.microsoftonline.com"
+BASE_URL = "https://graph.microsoft.com/v1.0"
+SCOPE = [
+    "User.Read.All",
+    "TeamMember.Read.All",
+    "ChannelMessage.Read.All",
+    "Chat.Read",
+    "Chat.ReadBasic",
+    "Calendars.Read",
+]
+
+
+class UserEndpointName(Enum):
+    PING = "Ping"
+
+
+URLS = {
+    UserEndpointName.PING.value: "{base_url}/me",
+}
+
+
+class NotFound(Exception):
+    """Internal exception class to handle 404s from the API that has a meaning, that collection
+    for specific object is empty.
+
+    It's not an exception for us, we just want to return [], and this exception class facilitates it.
+    """
+
+    pass
+
+
+class InternalServerError(Exception):
+    """Exception class to indicate that something went wrong on the server side."""
+
+    pass
+
+
+class PermissionsMissing(Exception):
+    """Exception class to notify that specific Application Permission is missing for the credentials used.
+    See: https://learn.microsoft.com/en-us/graph/permissions-reference
+    """
+
+    pass
+
+
+class TokenFetchFailed(Exception):
+    """Exception class to notify that connector was unable to fetch authentication token from either
+    Microsoft Teams Graph API.
+
+    Error message will indicate human-readable reason.
+    """
+
+    pass
+
+
+class GraphAPIToken:
+    """Class for handling access token for Microsoft Graph APIs"""
+
+    def __init__(self, tenant_id, client_id, client_secret, username, password):
+        """Initializer.
+
+        Args:
+            tenant_id (str): Azure AD Tenant Id
+            client_id (str): Azure App Client Id
+            client_secret (str): Azure App Client Secret Value
+            username (str): Username of the Azure account to fetch the access_token
+            password (str): Password of the Azure account to fetch the access_token"""
+
+        self.tenant_id = tenant_id
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.username = username
+        self.password = password
+
+        self._token_cache_with_username = CacheWithTimeout()
```

Comment:
There are 2 types of way to generate access token in `msal.ConfidentialClientApplication` to use graph api:
1.acquire_token_for_client
2.acquire_token_by_username_password
Some API's need `acquire_token_by_username_password` and some of them needs `acquire_token_for_client`. 
If we try to execute all the API's using clinet credentials only then it will throw precondition failed error to us:
`{"error":{"code":"PreconditionFailed","message":"Requested API is not supported in application-only context"}}`

For ping we need `acquire_token_by_username_password` that's why I kept this name here. We can change it to something else if you want.